crates/sui-core/src/execution_engine.rs

```diff
@@ -319,13 +323,24 @@ fn transfer_sui<S>(
         // This is necessary for the temporary store to know this new object is not unwrapped.
         let newly_generated_ids = tx_ctx.recreate_all_ids();
         temporary_store.set_create_object_ids(newly_generated_ids);
+        Some(amount)
     } else {
         // If amount is not specified, we simply transfer the entire coin object.
         // We don't want to increment the version number yet because latter gas charge will do it.
         object.transfer_without_version_change(recipient);
-    }
+        Coin::extract_balance_if_coin(&object)?
+    };
 
-    // TODO: Emit a new event type for this.
+    temporary_store.log_event(Event::TransferObject {
+        package_id: ObjectID::from(SUI_FRAMEWORK_ADDRESS),
+        transaction_module: Identifier::from(ident_str!("native")),
+        sender: tx_ctx.sender(),
+        recipient: Owner::AddressOwner(recipient),
+        object_id: object.id(),
+        version: object.version(),
+        type_: TransferType::Coin, // Should this be a separate type, like SuiCoin?
+        amount: transferred,
+    });
```

Comment:
I've updated the PR title.  Yes an event was not emitted before for `TransferSui` transactions, that was an oversight.

> most of this data is redundant with existing transaction data, yet we are adding it all to the core data store for what may be one of the most common types of events

So a couple of things here.... there were a bunch of discussions about the best place for this to go, some of it might be in the issue linked in the description, but the consensus was that we would put it in an event.   Yes, the object version is elsewhere in the effects, but not the transfer type and the amount.  The transfer type cannot be deduced from effects for Move transactions, for example, and the amount is hidden in the Coin object layout itself.  The destination of the transfer also is not directly derivable, though you could get it out of looking up the new object owner.

I agree that some of this data is redundant, and could be derived later, but this is really a larger discussion.   One consideration is that it is really important that events be part of the historical data record -- partners and users will want to look up events, especially this amount field -- and if the effects from the validator do not include the full events, then we have a discrepancy between the state of the validator and full nodes.   This discrepancy makes coming up with a full set of data for a checkpoint much more complicated.

I would not worry too much about redundant data and data storage costs right now.  The long term data storage will probably look nothing like it does today.